src/websocket/stream.py

```diff
@@ -1,113 +1,124 @@
+                    self.total_bytes += message
                     self.export_audio = self.concatenate_audio_with_crossfade(
                         self.export_audio, message
                     )
 
-                    if len(self.chunk_cache) >= self.final_treshold:
-                        self.logger.debug(
-                            f"NEW FINAL: length of chunk cache: {len(self.chunk_cache)}"
-                        )
-                        asyncio.ensure_future(
-                            self.transcribe_all_chunk_cache(
-                                websocket,
-                                self.chunk_cache,
-                                self.recently_added_chunk_cache,
+                    if (
+                        self.bytes_received_since_last_transcription >= self.partial_transcription_byte_threshold 
+                        or (time.time() - self.last_transcription_timestamp >= (self.partial_transcription_byte_threshold / BYTES_PER_SECOND))
+                    ):
+                        self.bytes_received_since_last_transcription = 0
+                        self.last_transcription_timestamp = time.time()
+
+                        # Ensure that no duplicate transcription jobs are running
+                        # This additonal check is needed as we are in nested async
+                        if len(self.transcription_tasks) < 1:
+                            self.logger.info(
+                                f"NEW PARTIAL: length of current window: {len(self.sliding_window)}"
                             )
-                        )
-                        self.recently_added_chunk_cache = b""
-                        self.chunk_cache = b""
-                        self.agreement.clear()
-
-                    if len(self.recently_added_chunk_cache) >= self.partial_treshold:
-                        self.logger.info(
-                            f"NEW PARTIAL: length of chunk cache: {len(self.recently_added_chunk_cache)}"
-                        )
-                        # Here we need to ensure that the transcribe_chunk_partial does not run longer than the length of the chunks.
-                        asyncio.ensure_future(
-                            self.transcribe_chunk_partial(
-                                websocket,
-                                self.chunk_cache,
-                                self.recently_added_chunk_cache,
+                            task = asyncio.create_task(
+                                self.transcribe_sliding_window(
+                                    websocket,
+                                    self.sliding_window
+                                ),
+                                name=f"transcription_task_stream_{self.id}"
                             )
+                            self.transcription_tasks.add(task)
+                            task.add_done_callback(self.transcription_tasks.discard)
```

Comment:
```suggestion
                           task.add_done_callback(lambda t: self.transcription_tasks.discard(t))
```

I don't understand, ask GPT

**Why This Fix?** (This is including below fix in finally)
add_done_callback(lambda t: self.transcription_tasks.discard(t)) ensures the task is removed only after it completes.
Using await asyncio.wait(self.transcription_tasks, return_when=asyncio.ALL_COMPLETED) ensures that all tasks finish before removing them.
self.transcription_tasks.difference_update(done) properly clears only completed tasks, avoiding race conditions.

Reply:
Not needed. This way of handling it is directly from the python docs and therefore should be the correct way

https://docs.python.org/3/library/asyncio-task.html#asyncio.create_task